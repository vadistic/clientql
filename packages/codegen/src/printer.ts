import { mergeExtensions, truthy } from '@graphql-clientgen/core'
import {
  ASTNode,
  buildASTSchema,
  DocumentNode,
  extendSchema,
  GraphQLSchema,
  Kind,
  parse,
} from 'graphql'
import { CodegenPrinterConfig, defaultConfig } from './config'
import {
  printArgument,
  printFieldDefinition,
  printInputValue,
  printNamedType,
  printObjectField,
  printSchemaDefinition,
  printType,
  printValue,
} from './type-reference'
import {
  printEnum,
  printInputObject,
  printInterface,
  printObject,
  printScalar,
  printUnion,
} from './type-system'

/**
 * Create printer - schema needs to be provided mainly for
 * - correct interfaces prefixing
 * - using extended interfaces
 *
 * can be skipped otherwise
 */

export const createCodegenPrinter = (
  config?: Partial<CodegenPrinterConfig>,
  schema?: GraphQLSchema | DocumentNode | string,
) => {
  let initalSchema: GraphQLSchema | undefined
  const initalConfig = { ...defaultConfig, ...config }

  /**
   * allow providing schema as doc | string | schema +  merge extensions when possible
   *
   * (in outer scope for performance)
   */

  const schemaOptions = {
    commentDescriptions: initalConfig.addDescription === 'comment',
    assumeValid: true,
    assumeValidSDL: true,
  }

  if (typeof schema === 'string') {
    initalSchema = buildASTSchema(mergeExtensions(parse(schema)), schemaOptions)
  } else if (schema && 'kind' in schema && schema.kind === Kind.DOCUMENT) {
    initalSchema = buildASTSchema(mergeExtensions(schema), schemaOptions)
  } else if (schema instanceof GraphQLSchema) {
    // no way to merge extensions :/
    initalSchema = schema
  }

  /**
   * codegen printer function
   */

  const printer = (
    ast: ASTNode | ASTNode[],
    configOverride?: Partial<CodegenPrinterConfig>,
  ) => {
    /*
     * dynamic extension of schema & config
     */

    let dynamicSchema = initalSchema
    const dynamicConfig = configOverride
      ? { ...initalConfig, ...configOverride }
      : initalConfig

    // try creating schema on the go when no initalSchema
    if (!initalSchema && !Array.isArray(ast) && ast.kind === Kind.DOCUMENT) {
      try {
        dynamicSchema = buildASTSchema(mergeExtensions(ast), schemaOptions)
      } catch (err) {
        /* noop */
      }
    }

    // merge doc into initalSchema if inital schema & provided ast was a doc
    if (initalSchema && !Array.isArray(ast) && ast.kind === Kind.DOCUMENT) {
      dynamicSchema = extendSchema(initalSchema, ast, schemaOptions)
    }

    /**
     * rescursive main loop
     */
    const printReducer = (node: ASTNode): string | null => {
      switch (node.kind) {
        /**
         * enter recursion
         */
        case Kind.DOCUMENT:
          return node.definitions.map(printReducer).join('\n\n')

        /**
         * type system definitions
         */
        case Kind.INPUT_OBJECT_TYPE_DEFINITION:
        case Kind.INPUT_OBJECT_TYPE_EXTENSION:
          return printInputObject(dynamicConfig, dynamicSchema)(node)
        case Kind.OBJECT_TYPE_DEFINITION:
        case Kind.OBJECT_TYPE_EXTENSION:
          return printObject(dynamicConfig, dynamicSchema)(node)
        case Kind.INTERFACE_TYPE_DEFINITION:
        case Kind.INTERFACE_TYPE_EXTENSION:
          return printInterface(dynamicConfig, dynamicSchema)(node)
        case Kind.UNION_TYPE_DEFINITION:
        case Kind.UNION_TYPE_EXTENSION:
          return printUnion(dynamicConfig, dynamicSchema)(node)
        case Kind.ENUM_TYPE_DEFINITION:
        case Kind.ENUM_TYPE_EXTENSION:
          return printEnum(dynamicConfig, dynamicSchema)(node)
        case Kind.SCALAR_TYPE_DEFINITION:
        case Kind.SCALAR_TYPE_EXTENSION:
          return printScalar(dynamicConfig, dynamicSchema)(node)

        /**
         * type reference
         */
        case Kind.INPUT_VALUE_DEFINITION:
          return printInputValue(dynamicConfig, dynamicSchema)(node)
        case Kind.FIELD_DEFINITION:
          return printFieldDefinition(dynamicConfig, dynamicSchema)(node)
        case Kind.NON_NULL_TYPE:
        case Kind.LIST_TYPE:
          return printType(dynamicConfig, dynamicSchema)(node)
        case Kind.NAMED_TYPE:
          return printNamedType(dynamicConfig, dynamicSchema)(node)
        case Kind.ARGUMENT:
          return printArgument(dynamicConfig, dynamicSchema)(node)

        /**
         * schema
         */
        case Kind.SCHEMA_DEFINITION:
        case Kind.SCHEMA_EXTENSION:
          return printSchemaDefinition(dynamicConfig, dynamicSchema)(node)

        /**
         * primitives
         */

        case Kind.INT:
        case Kind.FLOAT:
        case Kind.STRING:
        case Kind.BOOLEAN:
        case Kind.ENUM:
        case Kind.NULL:
        case Kind.LIST:
        case Kind.OBJECT:
        case Kind.VARIABLE:
          return printValue(dynamicConfig, dynamicSchema)(node)
        case Kind.OBJECT_FIELD:
          return printObjectField(dynamicConfig, dynamicSchema)(node)
        case Kind.NAME:
          return node.value
        default:
          return null
      }
    }

    // supports ast as arrays
    if (Array.isArray(ast)) {
      return ast
        .map(printReducer)
        .filter(truthy)
        .join('\n\n')
    }

    return printReducer(ast)
  }

  return printer
}

/**
 * printer with all default config & no schema
 */
export const codegenPrinter = createCodegenPrinter()
