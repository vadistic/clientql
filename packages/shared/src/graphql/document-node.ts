import { DefinitionNode, DocumentNode } from 'graphql'
import { Kind } from './kind'

export const wrapDocument = (...nodes: DefinitionNode[]): DocumentNode => ({
  kind: Kind.DOCUMENT,
  definitions: nodes
})

export const unwrapDocument = (doc: DocumentNode): DefinitionNode[] => [
  ...doc.definitions
]
