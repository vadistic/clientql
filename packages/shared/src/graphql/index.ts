export * from './kind'
export * from './document-node'
export * from './field-node'
