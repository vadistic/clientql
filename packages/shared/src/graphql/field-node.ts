import { ArgumentNode, FieldNode, SelectionNode } from 'graphql'
import { Kind } from './kind'

export interface CreateFieldProps {
  fieldname: string
  arguments?: ArgumentNode[]
  selections?: SelectionNode[]
}

export const createField = (props: CreateFieldProps): FieldNode => ({
  kind: Kind.FIELD,
  name: {
    kind: Kind.NAME,
    value: props.fieldname
  },
  arguments: props.arguments,
  selectionSet: props.selections
    ? {
        kind: Kind.SELECTION_SET,
        selections: props.selections
      }
    : undefined
})
