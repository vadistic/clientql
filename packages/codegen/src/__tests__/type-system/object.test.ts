import { PRISMA_TYPEDEFS } from '@graphql-clientgen/testing'
import { buildASTSchema, Kind } from 'graphql'
import gql from 'graphql-tag'
import { createCodegenPrinter } from '../../printer'

describe('printer > ' + Kind.OBJECT_TYPE_DEFINITION, () => {
  it('addFieldsAsFunction: true', () => {
    const fixture = gql`
      type MyType {
        prop: String!
        arr: [String!]!
        # ignores function
        fn(data: String!): String!
      }
    `

    const print = createCodegenPrinter({ addFieldAsFunction: true })

    expect(print(fixture)).toMatchInlineSnapshot(`
                        "export interface MyType {
                          prop: () => string
                          arr: () => string[]
                          fn: (args: { data: string }) => string
                        }"
                `)
  })

  it('addTypename: false', () => {
    const fixture = gql`
      type MyType {
        prop: String!
        arr: [String!]!
        # ignores function
        fn(data: String!): String!
      }
    `

    const print = createCodegenPrinter({ addTypename: false })

    expect(print(fixture)).toMatchInlineSnapshot(`
                        "export interface MyType {
                          prop: string
                          arr: string[]
                          fn: string
                        }"
                `)
  })

  it('useExtendedInterfaces: true', () => {
    const print = createCodegenPrinter({
      useExtendedInterfaces: true,
      useMapsForEnum: false,
    })

    const fixture = gql`
      interface Animal {
        sea: Boolean
        land: Boolean
      }

      type Dog implements Animal {
        sea: Boolean
        land: Boolean
        size: Size
      }

      enum Size {
        SMALL
        MEDIUM
        LARGE
      }
    `

    expect(print(fixture)).toMatchInlineSnapshot(`
      "export interface Animal {
        sea?: boolean | null
        land?: boolean | null
      }
      
      export interface Dog extends Animal {
        __typename: 'Dog'
        size?: Size | null
      }
      
      export enum Size {
        SMALL = 'SMALL',
        MEDIUM = 'MEDIUM',
        LARGE = 'LARGE',
      }"
    `)
  })

  it('useFieldArgumentsInterfaces: true', () => {
    const schema = buildASTSchema(PRISMA_TYPEDEFS)

    const print = createCodegenPrinter(
      { addFieldAsFunction: true, useFieldArgumentsInterface: true },
      PRISMA_TYPEDEFS,
    )

    const fixture = schema.getQueryType()!.astNode!

    expect(print(fixture)).toMatchSnapshot()
  })
})
