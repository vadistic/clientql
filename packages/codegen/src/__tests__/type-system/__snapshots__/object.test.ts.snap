// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printer > ObjectTypeDefinition useFieldArgumentsInterfaces: true 1`] = `
"export interface Query {
  board: (args?: QueryBoardArgs) => Board | null
  boards: (args: QueryBoardsArgs) => Array<Board | null>
  boardsConnection: (args: QueryBoardsConnectionArgs) => BoardConnection
  post: (args?: QueryPostArgs) => Post | null
  posts: (args: QueryPostsArgs) => Array<Post | null>
  postsConnection: (args: QueryPostsConnectionArgs) => PostConnection
  thread: (args?: QueryThreadArgs) => Thread | null
  threads: (args: QueryThreadsArgs) => Array<Thread | null>
  threadsConnection: (args: QueryThreadsConnectionArgs) => ThreadConnection
  user: (args?: QueryUserArgs) => User | null
  users: (args: QueryUsersArgs) => Array<User | null>
  usersConnection: (args: QueryUsersConnectionArgs) => UserConnection
  node: (args?: QueryNodeArgs) => Node | null
}

export interface QueryBoardArgs {
  where: BoardWhereUniqueInput
}

export interface QueryBoardsArgs {
  where?: BoardWhereInput | null
  orderBy?: BoardOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryBoardsConnectionArgs {
  where?: BoardWhereInput | null
  orderBy?: BoardOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryPostArgs {
  where: PostWhereUniqueInput
}

export interface QueryPostsArgs {
  where?: PostWhereInput | null
  orderBy?: PostOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryPostsConnectionArgs {
  where?: PostWhereInput | null
  orderBy?: PostOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryThreadArgs {
  where: ThreadWhereUniqueInput
}

export interface QueryThreadsArgs {
  where?: ThreadWhereInput | null
  orderBy?: ThreadOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryThreadsConnectionArgs {
  where?: ThreadWhereInput | null
  orderBy?: ThreadOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryUserArgs {
  where: UserWhereUniqueInput
}

export interface QueryUsersArgs {
  where?: UserWhereInput | null
  orderBy?: UserOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryUsersConnectionArgs {
  where?: UserWhereInput | null
  orderBy?: UserOrderByInput | null
  skip?: number | null
  after?: string | null
  before?: string | null
  first?: number | null
  last?: number | null
}

export interface QueryNodeArgs {
  id: string
}"
`;
